src/tunnel.cc

```diff
@@ -672,15 +781,40 @@ tunnelTimeout(const CommTimeoutCbParams &io)
     /* Temporary lock to protect our own feets (comm_close -> tunnelClientClosed -> Free) */
     CbcPointer<TunnelStateData> safetyLock(tunnelState);
 
-    tunnelState->client.closeIfOpen();
-    tunnelState->server.closeIfOpen();
+    tunnelState->closeConnections();
 }
 
 void
-TunnelStateData::Connection::closeIfOpen()
+TunnelStateData::closeConnections()
 {
-    if (Comm::IsConnOpen(conn))
+    if (Comm::IsConnOpen(server.conn))
+        server.conn->close();
+    if (Comm::IsConnOpen(client.conn))
+        client.conn->close();
+}
+
+void
+TunnelStateData::Connection::closeQuietly()
+{
+    assert(!writer);
+
+    if (Comm::IsConnOpen(conn)) {
+        Must(closer);
+        comm_remove_close_handler(conn->fd, closer);
+        closer = nullptr;
         conn->close();
+        conn = nullptr;
```

Comment:
Do you mean that it would be better to call just conn->close(), do not remove the handler (tunnelServerClosed()) so that it would perform all retry/cleanup (including retryOrBail())?
In this case, roughly, the call sequence would be:
`retryOrBail() ->conn->close()->tunnelServerClosed()->retryOrBail()->retry`
Currentlry, it is
`retryOrBail() ->closeQuietly()->retry`
, which is simpler. In other words, since we close the connection ourselves, we can eliminate the callback call and do all cleanup at once.

Reply:
I suspect that @yadij means that `conn` should not be set to nil after calling `comm->close()` in this "get rid of the connection" method. I disagree. The `conn` member represents tunnel's association with the current to-server connection, not the "last" to-server connection. Thus, the member should be cleared when the current connection is gone.

> The non-FD data stored there may be needed for logging, debugs or error display later.

The method caller has decided to get rid of the current to-server connection. If the caller wanted to preserve connection information, it could have done so (e.g., in an error or ALE object). Such preservation is not _this_ method business.

There are known serious problems with Comm::Connection API(*), but we are not going to solve those problems here. Fortunately, this PR can accomplish its goals without solving those API problems.

----

(*) including Connection use for storing things not associated with any existing connection and the fact that `isOpen(conn->fd) != IsOpen(conn)` for some Connections.

Reply:
@rousskov so where can you see the logging getting its information about the connection last tried when the Pointer to that data is nil?
 Also, how is retry going to re-open the connection when there is no record of where it was going?

Reply:
> where can you see the logging getting its information about the connection last tried

Access-logging code should get its information from ALE (and, where really necessary, HttpRequest and similar high-level legacy locations). Certainly not from TunnelStateData::Connection that it cannot even access because the TunnelStateData class is private to tunnel.cc!

> ... when the Pointer to that data is nil?

For logging code, it does not matter whether TunnelStateData::server.conn is nil. That code has no access to that pointer. And even if it did have access, that pointer is _not_ a pointer to "the connection last tried". It is the pointer to the "currently connected connection", as documented. 
Naturally, this pointer should be nil when there is no such connection. I see no other reasonable way to use Connection pointers in code that goes through multiple connections.

> how is retry going to re-open the connection when there is no record of where it was going?

If/when TunnelStateData supports reforwarding requests to the same destination, it will probably get that reforwarding address from the TunnelStateData::destinations collection. FwdState works that way already.